### PR TITLE
Batched Diagonal Operator

### PR DESCRIPTION
**Pull request by vpuri3**
fixes https://github.com/SciML/SciMLOperators.jl/issues/81

### CODE REVIEW

**Comment by vpuri3** [comment]
standard tests passing. now just gotta look at the `update_coefficients!` behavior with adjoints.

**Comment by vpuri3** [comment]
Lemme take a look at the update functionality in a separate PR.

**Comment by vpuri3** [comment]
@ChrisRackauckas this is done

File: docs/src/interface.md [hunk 1]
```diff
@@ -1,22 +1,35 @@
...
 These are the formal properties that an `AbstractSciMLOperator` should obey
 for it to work in the solvers.
 
-## AbstractDiffEqOperator Interface Description
+1. An `AbstractSciMLOperator` represents a linear or nonlinear operator with input/output
+   being `AbstractArray`s. Specifically, a SciMLOperator, `L`, of size `(M,N)` accepts
+   input argument `u` with leading length `N`, i.e. `size(u, 1) == N`, and returns an
+   `AbstractArray` of the same dimension with leading length `M`, i.e. `size(L * u, 1) == M`.
+   Internally, `L` lazily reshapes `u` to a matrix of size `(N, length(u) \div N)` and
+   independently acts on the column-vectors. The reshape operation is skipped for
```
Comment (ChrisRackauckas) [inline docs/src/interface.md:13]: No, a counter example of this is derivative operations and convolutions. Columns are not independent. It's a linear operator on the whole `vec(u)`.

File: docs/src/interface.md [hunk 2]
```diff
@@ -1,22 +1,35 @@
...
+   `AbstractVecOrMat` arguments.
+2. SciMLOperators can be applied to an `AbstractArray` via overloaded `Base.*`, or
+   the in-place `LinearAlgebra.mul!`. Additionally, operators are allowed to be time,
+   or parameter dependent. The state of a SciMLOperator can be updated by calling
+   the mutating function `update_coefficients!(L, u, p, t)` where `p` representes
+   parameters, and `t`, time.  Calling a SciMLOperator as `L(du, u, p, t)` or out-of-place
+   `L(u, p, t)` will automatically update the state of `L` before applying it to `u`.
```
Comment (ChrisRackauckas) [inline docs/src/interface.md:20]: or simply, `L(u,p,t)` is the operation `L(u,p,t) * u` as is common in the definition of semilinear ODE systems.

File: docs/src/interface.md [hunk 3]
```diff
@@ -1,22 +1,35 @@
...
+3. To support the update functionality, we have lazily implemented a comprehensive operator
+   algebra. That means a user can add, subtract, scale, compose and invert SciMLOperators,
+   and the state of the resultant operator would be updated as expected upon calling
+   `L(du, u, p, t)` or `L(u, p, t)` so long as an update function is provided for the
+   component operators.
+4. Out of place `L = update_coefficients(L, u, p, t)`
```
Comment (ChrisRackauckas) [inline docs/src/interface.md:26]: incomplete sentence.

File: docs/src/interface.md [hunk 4]
```diff
@@ -28,6 +41,7 @@ for it to work in the solvers.
 8. Optional: factorizations. `ldiv!`, `factorize` et. al. This is only required
    for algorithms which use the factorization of the operator (Crank-Nicolson),
    and only for when the default linear solve is used.
+1. `AbstractSciMLLinearOperator <: AbstractSciMLOperator`
```
Comment (ChrisRackauckas) [inline docs/src/interface.md:44]: 1 is last?

File: docs/src/interface.md [hunk 5]
```diff
@@ -1,22 +1,35 @@
...
-1. Function call and multiplication: `L(du,u,p,t)` for inplace and `du = L(u,p,t)` for
-   out-of-place, meaning `L*u` and `mul!`.
-2. If the operator is not a constant, update it with `(u,p,t)`. A mutating form, i.e.
-   `update_coefficients!(A,u,p,t)` that changes the internal coefficients, and a
-   out-of-place form `B = update_coefficients(A,u,p,t)`.
-3. `isconstant(A)` trait for whether the operator is constant or not.
+## AbstractSciMLOperator Interface Description
 
-## AbstractDiffEqLinearOperator Interface Description
+3. `isconstant(A)` trait for whether the operator is constant or not.
```
Comment (ChrisRackauckas) [inline docs/src/interface.md:31]: 3 is first?

File: src/batch.jl [hunk 6]
```diff
@@ -0,0 +1,110 @@
+#
+struct BatchedDiagonalOperator{T,D,F} <: AbstractSciMLLinearOperator{T}
```
Comment (ChrisRackauckas) [inline src/batch.jl:2]: docstring is missing.

File: src/batch.jl [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    end
+end
+
+"""
+When `diag` is a multidimensional array, `L = DiagonalOperator(diag, ...)` forms
+an operator of size `(N, N)` where `N = size(diag, 1)` is the leading length of `diag`.
+`L` then is the elementwise-scaling operation on arrays of `length(u) = length(diag)`
+with leading length `size(u, 1) = N`. Note that `L` in this case doesn't act independently
+on coumn-vectors of its input.
+"""
```
Comment (ChrisRackauckas) [inline src/batch.jl:26]: please use the standard docstring form. This is a comment not a docstring.

File: src/matrix.jl [hunk 8]
```diff
@@ -78,13 +78,25 @@ LinearAlgebra.mul!(v::AbstractVecOrMat, L::MatrixOperator, u::AbstractVecOrMat,
 LinearAlgebra.ldiv!(v::AbstractVecOrMat, L::MatrixOperator, u::AbstractVecOrMat) = ldiv!(v, L.A, u)
 LinearAlgebra.ldiv!(L::MatrixOperator, u::AbstractVecOrMat) = ldiv!(L.A, u)
 
-""" Diagonal Operator """
-function DiagonalOperator(u::AbstractVector; update_func=DEFAULT_UPDATE_FUNC)
+"""
+    DiagonalOperator(diag, [; update_func])
```
Comment (ChrisRackauckas) [inline src/matrix.jl:82]:
```suggestion
```julia
DiagonalOperator(diag, [; update_func])
```
```

File: src/matrix.jl [hunk 9]
```diff
@@ -78,13 +78,25 @@ LinearAlgebra.mul!(v::AbstractVecOrMat, L::MatrixOperator, u::AbstractVecOrMat,
...
+
+Represents a time-dependent elementwise scaling (diagonal-scalign) operation.
```
Comment (ChrisRackauckas) [inline src/matrix.jl:84]:
```suggestion
Represents a time-dependent elementwise scaling (diagonal-scaling) operation.
```

File: src/matrix.jl [hunk 10]
```diff
@@ -78,13 +78,25 @@ LinearAlgebra.mul!(v::AbstractVecOrMat, L::MatrixOperator, u::AbstractVecOrMat,
...
+The update function is called by `update_coefficients!` and is assumed to have
+the following signature:
+
+    update_func(diag::AbstractVector,u,p,t) -> [modifies diag]
+
+When `diag` is an `AbstractVector` of length N, `L=DiagonalOpeator(diag, ...)`
+can be applied to `AbstractArray`s with `size(u, 1) == N`. Each column of the `u`
+will be scaled by `diag`, as in `LinearAlgebra.Diagonal(diag) * u`.
+"""
```
Comment (ChrisRackauckas) [inline src/matrix.jl:93]:
```julia

## Keyword Arguments

```

etc.

**Comment by ChrisRackauckas** [comment]
See what's recommended on types and functions:

https://github.com/SciML/SciMLStyle#documentation

**Comment by vpuri3** [comment]
cool, thanks for the pointers. ill get it up and running

**Comment by vpuri3** [comment]
ok done

**Comment by ChrisRackauckas** [comment]
It's still not documented appropriately. This whole repo will need a docs overhaul to fit the style. We can do that in a future PR though.

**Pull request merged**